### PR TITLE
feat: Augmentations on the GPU

### PR DESCRIPTION
**Pull request by mogwai**
There had been many changes since my last attempt #508 so I decided to start from scratch.

We realised the only way to get lightning to get the augmentations on the GPU was for them to be part of the model. That way, regardless of the multi GPU training method being used, anywhere the model is, it will have access to use the augmentations. This unfortunately does clog up the model some what but is necessary at the moment to get the augmentations working on each device. There may be another way to do this in the future.

This particular feature is hard to test with our current setup as we haven't got a good way of running tests on the GPU, perhaps this is something that we really need soon. I'd be happy to have a look into it.

### CODE REVIEW

**Comment by hbredin** [comment]
Thanks. Any idea why CI checks haven't completed yet?

**Comment by mogwai** [comment]
My PR's don't get CI checked and I have no idea why. Maybe Github thinks I'm DDOS'ing them. Regardless you probably need to run this locally yourself to make sure that it works on multiple GPUS's and that nothing odd is happening.

File: pyannote/audio/augmentation/registry.py [hunk 1]
```diff
@@ -94,41 +89,39 @@ def forward(self, spectrograms):
 
     """
 
-    wrapped_augmentation = wrap_augmentation(augmentation, module, when=when)
```
Comment (hbredin) [inline pyannote/audio/augmentation/registry.py:97]: What are the motivation for getting rid of this wrapper? 

I thought it was needed to pass `sample_rate` to `torch-audiomentations` transform `.forward`.

Comment (mogwai) [reply pyannote/audio/augmentation/registry.py:97]: This isn't needed anymore, you can specify the sample rate in the init

Comment (hbredin) [reply pyannote/audio/augmentation/registry.py:97]: I understand but that also means that you need to repeat `sample_rate` multiple times (once for the model, once for the augmentation). With the wrapper, you only have to do it once for the model -- ensuring sample rate consistency automatically.

Comment (mogwai) [reply pyannote/audio/augmentation/registry.py:97]: That is true, the main reason I pulled it out was because of cyclic dependencies when importing the registeration method in the core.task/model files. I'll see if I can get around that, otherwise we could iterate of the modules in task.augmentation and set the sample rate there.

Comment (mogwai) [reply pyannote/audio/augmentation/registry.py:97]: I wasn't able to resolve the cyclic dependencies but I created a method that checks the augmentations in the setup and gives them a sample rate if they don't have one.

File: pyannote/audio/core/task.py [hunk 2]
```diff
@@ -172,8 +174,11 @@ def __init__(
         num_workers: int = None,
         pin_memory: bool = False,
         augmentation: BaseWaveformTransform = None,
+        gpu_transforms: bool = True,
     ):
         super().__init__()
+        # gpu transforms
+        self.gpu_transforms = gpu_transforms and torch.cuda.is_available()
```
Comment (hbredin) [inline pyannote/audio/core/task.py:181]: As a follow-up of the above comment (line 177), we no longer have to check whether `cuda` is available.

File: pyannote/audio/core/task.py [hunk 3]
```diff
@@ -232,14 +237,27 @@ def setup(self, stage=None):
...
         """
-        pass
 
     def setup_loss_func(self):
         pass
 
     def setup_validation_metric(self):
         pass
 
+    def _set_augmentation_sample_rate(self, augmentation: BaseWaveformTransform):
```
Comment (hbredin) [inline pyannote/audio/core/task.py:247]: This feels like a hack -- I'd rather not have this kind of logic within `Task` class.
I'd rather not touch `augmentation` internals from within `Task`.

Furthermore, what if we want to support another augmentation library later? 
This will be a nighmare to maintain...

Hence my original proposal to let the augmentation registry handle this.
We'd only have to write another `@singledispatch` entry in `registry`.

File: pyannote/audio/core/task.py [hunk 4]
```diff
@@ -257,7 +275,7 @@ def train__len__(self):
 
     def collate_fn(self, batch):
         collated_batch = default_collate(batch)
-        if self.augmentation is not None:
+        if self.augmentation is not None and not self.gpu_transforms:
```
Comment (hbredin) [inline pyannote/audio/core/task.py:278]: This would have to be renamed once we switch to `in model` vs. `in dataloader` convention.

File: pyannote/audio/core/model.py [hunk 5]
```diff
@@ -378,6 +378,8 @@ def setup(self, stage=None):
             self.task.setup_validation_metric()
             # this is to make sure introspection is performed here, once and for all
             _ = self.introspection
+            # Put the augmentations on the gpu if we can by registering them on the model
+            self.task._setup_augmentations()
```
Comment (hbredin) [inline pyannote/audio/core/model.py:382]: Why the `_` prefix? Let's be consistent with the other `setup_xxx` methods.

Comment (hbredin) [reply pyannote/audio/core/model.py:382]: Also, I think we should `unregister` the augmentation before checkpoint the model.
Otherwise, there might be problems when loading them back? 

Did you test loading a model that was trained with `in model` transforms?
Does this work?

**Review by hbredin — COMMENTED** [review]
I finally found some time to look at this.

To be honest, I am less and less convinced about the advantage of running augmentation on the GPU when we can let multiple workers do that on CPU in parallel.

**Comment by mogwai** [comment]
> To be honest, I am less and less convinced about the advantage of running augmentation on the GPU when we can let multiple workers do that on CPU in parallel.

If it is faster that would be an advantage, so testing that we can further make a decision on this. I can do this on one a system with one GPU but I haven't got access to a machine with two. Are you concerned about the loss of GPU memory? Are there any other concerns you have?

**Comment by hbredin** [comment]
> > To be honest, I am less and less convinced about the advantage of running augmentation on the GPU when we can let multiple workers do that on CPU in parallel.
> 
> If it is faster that would be an advantage, so testing that we can further make a decision on this. I can do this on one a system with one GPU but I haven't got access to a machine with two. 

We should indeed benchmark training speed.

> Are you concerned about the loss of GPU memory? Are there any other concerns you have?

Memory might be an issue. I am also concerned about the fact that some augmentation techniques don't have a GPU-compatible implementation (e.g. encoding audio with several audio codecs). This would result in situations  where the model would have some parts on CPU and some other parts on GPU. In particular, since the number of GPUs is much smaller that the number of CPUs, this would result in a massive speed reduction. 

I am also concerned about how to make sure that augmentation layers are removed from the model before checkpointing -- this is important in particular for augmentation layers such as background noise addition that relies on a local collection of files. They would fail miserably when run on another computer.

**Comment by hbredin** [comment]
Closing. Might reopen at some point if the need for faster augmentation appears.

**Comment by mogwai** [comment]
Another downside to this approach, is that it's more important to have a larger batch size. If these augmentations take up a lot of space on the gpu memory, it might not be worth it.
